### PR TITLE
Import feature flag list feature

### PR DESCRIPTION
**Pull request by Yagnik56**
This PR contains an import feature flag list feature with validation and import endpoints. 

Here are a few images of the same.
image
image
image

There are 3 lists json files available in zack's comment for testing with the updated format.

Some Validation user has to keep in mind:
The list's context and feature flag's context must match; all these lists have Mathia as context. 
The feature flag ID isn't needed. It will be removed from export and where the list is getting imported that feature flag's ID will be added during import from the backend. 
For the listType as a segment, the subsegment it refers to should exist with the same context as the feature flag we are importing that list in. 
The group type has to be valid for its context with the list type available for it's context.

### CODE REVIEW

File: frontend/projects/upgrade/src/app/features/dashboard/feature-flags/pages/feature-flag-details-page/feature-flag-details-page-content/feature-flag-exclusions-section-card/feature-flag-exclusions-section-card.component.ts [hunk 1]
```diff
@@ -63,8 +64,21 @@ export class FeatureFlagExclusionsSectionCardComponent {
     this.dialogService.openAddExcludeListModal(appContext, flagId);
   }
 
-  onMenuButtonItemClick(event) {
+  onMenuButtonItemClick(event, flag: FeatureFlag) {
     console.log('Menu Button Item Clicked:', event);
```
Comment (zackcl) [inline frontend/projects/upgrade/src/app/features/dashboard/feature-flags/pages/feature-flag-details-page/feature-flag-details-page-content/feature-flag-exclusions-section-card/feature-flag-exclusions-section-card.component.ts:68]: Let's remove the logging which was added for testing.

**Comment by zackcl** [comment]
@Yagnik56 Could you please share a working FF list JSON file (or content)?

**Comment by zackcl** [comment]
Thanks for sharing the files. I tried importing the files you shared, but they all show up as "Incompatible."
Screenshot 2024-10-22 at 1 34 36 AM

Could it be because the file doesn't match the feature flag I'm trying to import it into?

```
{"createdAt":"2024-10-03T08:22:14.136Z","updatedAt":"2024-10-03T08:22:23.888Z","versionNumber":2,"segmentId":"306aeaac-0f58-4d28-8c6a-918430dc7218","featureFlagId":"ffb4b749-3127-44ff-9aae-e011dacbad73","enabled":true,"listType":"Individual","segment":{"createdAt":"2024-10-03T08:22:14.136Z","updatedAt":"2024-10-03T08:22:14.136Z","versionNumber":1,"id":"306aeaac-0f58-4d28-8c6a-918430dc7218","name":"bfdb","description":"","context":"mathia","type":"private","individualForSegment":[{"createdAt":"2024-10-03T08:22:23.888Z","updatedAt":"2024-10-03T08:22:23.888Z","versionNumber":1,"segmentId":"306aeaac-0f58-4d28-8c6a-918430dc7218","userId":"bfbfd"}],"groupForSegment":[],"subSegments":[]}}
```

By the way, I'm not sure why we need to include the feature-flag-specific properties in the List file. These files should be exported from and imported into different feature flags (Include and Exclude Lists) and segments (Lists) tables.

Could we simply include list-specific properties in the List file? I think the properties might differ depending on the "listType."

**Comment by Yagnik56** [comment]
The list's context and feature flag must match; all these lists have mathia as context. The feature flag ID isn't needed. It will be removed from export and where the list is getting imported that feature flag's ID will be added during import from the backend. For the listType segment, the subsegment it refers to should exist with the same context as the feature flag we are importing that list in. For a group, the group type has to be valid for its context.

File: frontend/projects/upgrade/src/app/shared/services/common-dialog.service.ts [hunk 2]
```diff
@@ -357,7 +359,43 @@ export class DialogService {
...
+      params: {
+        listType: null,
+        flagId: null,
+      },
+    };
+    return this.openImportModal(commonModalConfig);
+  }
+
+  openImportFeatureFlagIncludeListModal(flagId: string) {
+    const commonModalConfig: CommonModalConfig<ImportListParams> = {
```
Comment (VivekFitkariwala) [inline frontend/projects/upgrade/src/app/shared/services/common-dialog.service.ts:371]: This can be refactored to be a common object

**Review by VivekFitkariwala — CHANGES_REQUESTED** [review]

**Comment by zackcl** [comment]
@Yagnik56 Could you please reshare a JSON file content with minimal information possible? I think you mentioned many of the properties (e.g., `featureFlagId`) won't be included in the JSON file when exported, right? It would be helpful to see the file content for discussion tomorrow (we could discuss whether we want to include the `context ` in the file or not).

**Comment by Yagnik56** [comment]
@zackcl I updated the `json` files in the description and removed the data that the backend won't send in export. You can have a look.

**Comment by zackcl** [comment]
Here are the file contents for reference:
----------
Individual_list.json:
```
{
   "enabled":true,
   "listType":"Individual",
   "segment":{
      "name":"bfdb",
      "description":"",
      "context":"mathia",
      "type":"private",
      "individualForSegment":[
         {
            "createdAt":"2024-10-03T08:22:23.888Z",
            "updatedAt":"2024-10-03T08:22:23.888Z",
            "versionNumber":1,
            "segmentId":"306aeaac-0f58-4d28-8c6a-918430dc7218",
            "userId":"bfbfd"
         }
      ],
      "groupForSegment":[
         
      ],
      "subSegments":[
         
      ]
   }
}
```

Group_list.json:
```
{
   "enabled":false,
   "listType":"group1",
   "segment":{
      "name":"fbdfd",
      "description":"",
      "context":"mathia",
      "type":"private",
      "individualForSegment":[
         
      ],
      "groupForSegment":[
         {
            "createdAt":"2024-10-03T08:22:19.392Z",
            "updatedAt":"2024-10-03T08:22:19.392Z",
            "versionNumber":1,
            "segmentId":"1be939c9-afcb-49ac-bc77-28bf8fc2126f",
            "groupId":"fbdf",
            "type":"group1"
         }
      ],
      "subSegments":[
         
      ]
   }
}
```

Segment_list.json:

```
{
   "enabled":true,
   "listType":"Segment",
   "segment":{
      "name":"seg1",
      "description":"2",
      "context":"mathia",
      "type":"private",
      "individualForSegment":[
         
      ],
      "groupForSegment":[
         
      ],
      "subSegments":[
         {
            "createdAt":"2024-10-01T08:59:24.244Z",
            "updatedAt":"2024-10-01T09:10:14.316Z",
            "versionNumber":4,
            "id":"a78e3c7d-8d55-438a-afd3-c9539fbb870a",
            "name":"seg1",
            "description":"2",
            "context":"mathia",
            "type":"public"
         }
      ]
   }
}
```

File: backend/packages/Upgrade/src/api/services/FeatureFlagService.ts [hunk 3]
```diff
@@ -1092,4 +1105,188 @@ export class FeatureFlagService {
...
+          content: null,
+        };
+      }
+    });
+
+    const featureFlag = await this.findOne(featureFlagId, logger);
+
+    const validationErrors = await Promise.allSettled(
+      parsedFeatureFlagLists.map(async (parsedFile) => {
+        if (!featureFlag) {
```
Comment (VivekFitkariwala) [inline backend/packages/Upgrade/src/api/services/FeatureFlagService.ts:1205]: Write a single if statement 
```suggestion
        if (!featureFlag || (!parsedFile.content))
```

File: backend/packages/Upgrade/src/api/services/FeatureFlagService.ts [hunk 4]
```diff
@@ -1092,4 +1105,188 @@ export class FeatureFlagService {
...
+        }
+
+        if (!parsedFile.content) {
+          return {
+            fileName: parsedFile.fileName,
+            compatibilityType: FF_COMPATIBILITY_TYPE.INCOMPATIBLE,
+          };
+        }
+
+        const error = await this.validateImportFeatureFlagList(parsedFile.fileName, featureFlag, parsedFile.content);
```
Comment (VivekFitkariwala) [inline backend/packages/Upgrade/src/api/services/FeatureFlagService.ts:1219]: Return the function itself

File: backend/packages/Upgrade/src/api/services/FeatureFlagService.ts [hunk 5]
```diff
@@ -1092,4 +1105,188 @@ export class FeatureFlagService {
...
+            ...list,
+            enabled: false,
+            flagId: featureFlagId,
+            segment: { ...list.segment, id: uuid() },
+          };
+
+          listDocs.push(listDoc);
+        }
+
+        if (listType === FEATURE_FLAG_PARTICIPANT_LIST_KEY.INCLUDE) {
```
Comment (VivekFitkariwala) [inline backend/packages/Upgrade/src/api/services/FeatureFlagService.ts:1150]: You don't need the if else case

File: frontend/projects/upgrade/src/app/shared/services/common-dialog.service.ts [hunk 6]
```diff
@@ -6,11 +6,13 @@ import { ImportFeatureFlagModalComponent } from '../../features/dashboard/featur
 import { UpsertFeatureFlagModalComponent } from '../../features/dashboard/feature-flags/modals/upsert-feature-flag-modal/upsert-feature-flag-modal.component';
 import { UpsertPrivateSegmentListModalComponent } from '../../features/dashboard/segments/modals/upsert-private-segment-list-modal/upsert-private-segment-list-modal.component';
 import {
+  ImportListParams,
```
Comment (VivekFitkariwala) [inline frontend/projects/upgrade/src/app/shared/services/common-dialog.service.ts:9]: Remove this

File: frontend/projects/upgrade/src/app/shared/services/common-dialog.service.ts [hunk 7]
```diff
@@ -352,11 +354,27 @@ export class DialogService {
...
+
+  openImportFeatureFlagIncludeListModal(flagId: string) {
+    return this.openImportModal('Import List', FEATURE_FLAG_PARTICIPANT_LIST_KEY.INCLUDE, flagId);
+  }
+
+  openImportFeatureFlagExcludeListModal(flagId: string) {
+    return this.openImportModal('Import List', FEATURE_FLAG_PARTICIPANT_LIST_KEY.EXCLUDE, flagId);
+  }
+
+  openImportModal(title, listType, flagId) {
```
Comment (VivekFitkariwala) [inline frontend/projects/upgrade/src/app/shared/services/common-dialog.service.ts:368]: Add types here

**Review by VivekFitkariwala — CHANGES_REQUESTED** [review]
Small changes.

**Review by VivekFitkariwala — APPROVED** [review]

**Review by VivekFitkariwala — APPROVED** [review]

**Pull request merged**